File: lua/05-functions/higher_order.lua
-- A higher-order function: it receives a function as an argument
local function apply(fn, value)
    return fn(value)
end

local function square(n)
    return n * n
end

print(apply(square, 6))

-- Anonymous functions can be passed inline
print(apply(function(n) return n + 1 end, 41))

-- A closure: make_counter returns a function that remembers `count`
local function make_counter()
    local count = 0
    return function()
        count = count + 1
        return count
    end
end

local next_id = make_counter()
print(next_id())
print(next_id())
print(next_id())

-- Each counter keeps its own independent state
local other = make_counter()
print(other())

File: perl/05-functions/higher_order.pl
#!/usr/bin/env perl
use strict;
use warnings;
use v5.10;

# A code reference: an anonymous subroutine stored in a scalar
my $square = sub {
    my ($n) = @_;
    return $n * $n;
};

say "square(6) = " . $square->(6);

# Pass a code reference to another subroutine
sub apply_twice {
    my ($func, $value) = @_;
    return $func->($func->($value));
}

say "apply_twice(square, 2) = " . apply_twice($square, 2);

# A closure remembers the lexical environment where it was created
sub make_counter {
    my $count = 0;
    return sub { return ++$count; };
}

my $counter = make_counter();
say "Counter: " . $counter->();
say "Counter: " . $counter->();
say "Counter: " . $counter->();

# Built-in higher-order functions: map and grep
my @numbers = (1, 2, 3, 4, 5, 6);
my @doubled = map  { $_ * 2 } @numbers;
my @evens   = grep { $_ % 2 == 0 } @numbers;

say "Doubled: @doubled";
say "Evens: @evens";
